### PR TITLE
Add Foundry Metadata Option to Native Token Creation

### PR DESCRIPTION
**Pull request by thibault-martinez**
Addresses https://github.com/iotaledger/wallet.rs/issues/988

## Change checklist

Tick the boxes that are relevant to your changes, and delete any items that are not.

- [ ] I have followed the contribution guidelines for this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that new and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/account/operations/transfer/high_level/mint_native_token.rs [hunk 1]
```diff
@@ -46,6 +46,9 @@ pub struct NativeTokenOptions {
     /// Maximum supply
     #[serde(rename = "maximumSupply")]
     pub maximum_supply: U256,
+    /// Foundry metadata
+    #[serde(rename = "foundryMetadata")]
+    pub foundry_metadata: Vec<u8>,
```
Comment (thibault-martinez) [inline src/account/operations/transfer/high_level/mint_native_token.rs]: Need to make that an option

File: src/account/operations/transfer/high_level/mint_native_token.rs [hunk 2]
```diff
@@ -152,9 +155,9 @@ impl AccountHandle {
             )?])?;
 
             let outputs = vec![
-                Output::Alias(new_alias_output_builder.finish()?),
-                Output::Foundry(
-                    FoundryOutputBuilder::new_with_amount(
+                new_alias_output_builder.finish_output()?,
+                {
+                    let mut foundry_builder = FoundryOutputBuilder::new_with_amount(
                         minimum_storage_deposit_foundry(&byte_cost_config)?,
```
Comment (Thoralf-M) [inline src/account/operations/transfer/high_level/mint_native_token.rs:162]: I think we need to change this, if there is `foundry_metadata`
And what I'm also not sure is if we should increase the storage deposit so the foundry can mint native tokens without further inputs (add the required storage deposit amount for one NativeTokens)

File: src/account/operations/transfer/high_level/mint_native_token.rs [hunk 3]
```diff
@@ -46,6 +46,9 @@ pub struct NativeTokenOptions {
     /// Maximum supply
     #[serde(rename = "maximumSupply")]
     pub maximum_supply: U256,
+    /// Foundry metadata
+    #[serde(rename = "foundryMetadata")]
+    pub foundry_metadata: Option<Vec<u8>>,
 }
 
 impl AccountHandle {
```
Comment (Thoralf-M) [inline src/account/operations/transfer/high_level/mint_native_token.rs:54]: Can't comment at line 67, but the example there needs to be updated with the foundry_metadata field

Comment (thibault-martinez) [reply src/account/operations/transfer/high_level/mint_native_token.rs:54]: Right, it's ignored so I missed it. Fixed.

File: src/account/operations/transfer/high_level/mint_native_token.rs (hunk 2 above)
Comment (thibault-martinez) [reply src/account/operations/transfer/high_level/mint_native_token.rs:162]: Regarding the first sentence you are right. But in https://github.com/iotaledger/wallet.rs/pull/1007 I'm removing `minimum_storage_deposit_foundry` and using `new_with_minimum_storage_deposit` so it should take care of the metadata to compute the appropriate amount. I'm not sure what you mean with the second sentence though, can you elaborate?

Comment (Thoralf-M) [reply src/account/operations/transfer/high_level/mint_native_token.rs:162]: For the minted native tokens we currently add a new basic output which will get them, but if the foundry would have a larger amount, we could also add the native tokens to the foundry output directly, that way we have one output less (and if we don't mint any native tokens in this call, we could the next time use only the alias and foundry output as input and mint the native tokens to the foundry output)

Comment (Thoralf-M) [reply src/account/operations/transfer/high_level/mint_native_token.rs:162]: Just a possible optimization, not really required

**Review by Thoralf-M — APPROVED** [review]

File: src/account/operations/transfer/high_level/mint_native_token.rs (hunk 2 above)
Comment (thibault-martinez) [reply src/account/operations/transfer/high_level/mint_native_token.rs:162]: Well at least the first sentence is an actual mistake

Comment (Thoralf-M) [reply src/account/operations/transfer/high_level/mint_native_token.rs:162]: Yes, but if it's fixed in https://github.com/iotaledger/wallet.rs/pull/1007 it should be fine

**Pull request merged**